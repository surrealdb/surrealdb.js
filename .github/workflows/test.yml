name: test

on:
  push:
  pull_request:

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
      - name: Setup node.js
        uses: actions/setup-node@v3
        with:
          node-version: 18
      - name: Setup deno
        uses: denoland/setup-deno@v1
        with:
          deno-version: v1.x
      - name: Setup bun
        uses: antongolub/action-setup-bun@v1
        with:
          bun-version: 0.1.13

      - uses: actions/checkout@v3

      - name: Build NPM package
        run: deno task build

      - name: Start SurrealDB
        run: sh ./test/e2e/scripts/docker-restart.sh

      - name: Run Deno
        run: deno test -A --trace-ops ./test/e2e/deno.js
      - name: Run Node
        run: node ./test/e2e/node.js
      - name: Run Bun
        run: bun run ./test/e2e/bun.js

      - name: Stop SurrealDB
        run:  sh ./test/e2e/scripts/docker-stop.sh
