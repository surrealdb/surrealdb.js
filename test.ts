import { RecordId } from "./src/library/data/recordid.ts";
import { Duration } from "./src/library/data/duration.ts";
import { UUID } from "./src/library/data/uuid.ts";
import { Decimal } from "./src/library/data/decimal.ts";
import { Surreal } from "./src/index.ts";
import { encode } from "./src/library/data/cbor.ts";
import { GeometryCollection, GeometryLine, GeometryPoint } from "./src/library/data/geometry.ts";
// import { recordId } from "./src/library/orm/types.ts";

const surreal = new Surreal();
await surreal.connect('ws://127.0.0.1:8000/rpc');
await surreal.use({
	namespace: 'test',
	database: 'test',
});

// console.log(
// 	await surreal.query<[unknown[]]>(
// 		"[$d, $r, $c, $u, $dur, $none, NONE, $a, <string> $a, $nan, $geo]",
// 		{
// 			d: new Date(),
// 			r: new RecordId('person', 'tobie'),
// 			c: new RecordId('recording', [
// 				new Date(),
// 				'London',
// 				new Uuid(),
// 				{
// 					temparature: new Decimal("10.00003")
// 				}
// 			]),
// 			u: new Uuid(),
// 			dur: new Duration('1d20m'),
// 			none: undefined,
// 			a: 9223372036854775807n,
// 			nan: NaN,
// 			geo: new GeometryCollection([
// 				new GeometryPoint([1, 2]),
// 				new GeometryLine([ new GeometryPoint([1, 2]), new GeometryPoint([3, 4]) ])
// 			])
// 		}
// 	)
// )

const uuid = new Duration("2w6d100ms");
console.log(
	456,
	await surreal.query<[UUID, string]>(
		"$uuid; <string> $uuid",
		{
			uuid
		}
	).then(([u, a]) => [u.toString(), a])
)

// function bufferToHex (buffer: ArrayBuffer) {
//     return [...new Uint8Array (buffer)]
//         .map (b => b.toString (16).padStart (2, "0"))
//         .join ("");
// }

// console.log(bufferToHex(encode(new RecordId('recording', [
// 	new Date(),
// 	'London',
// 	new Uuid(),
// 	{
// 		temparature: new Decimal("10.00003")
// 	}
// ]))))

// const rid = recordId('test');
// console.log(rid.parse(new RecordId('tesst', 123) ))

// console.log()
