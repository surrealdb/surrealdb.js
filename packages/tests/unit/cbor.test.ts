import { describe, expect, test } from "bun:test";
import {
	CborFillMissing,
	CborInvalidMajorError,
	CborPartialDisabled,
	CborRangeError,
	Gap,
	POW_2_64,
	decode,
	encode,
} from "@surrealdb/cbor";

describe("cbor", () => {
	test("encode basic types", () => {
		expect(encode(123)).toMatchSnapshot("positive integer");
		expect(encode(-123)).toMatchSnapshot("negative integer");
		expect(encode(123.456)).toMatchSnapshot("positive float");
		expect(encode(-123.456)).toMatchSnapshot("negative float");
		expect(encode(POW_2_64 - 1n)).toMatchSnapshot("positive bigint");
		expect(encode(-POW_2_64)).toMatchSnapshot("negative bigint");
		expect(encode("Hello \nWorld!")).toMatchSnapshot("string");
		expect(encode(undefined)).toMatchSnapshot("undefined");
		expect(encode(null)).toMatchSnapshot("null");
		expect(encode(true)).toMatchSnapshot("true");
		expect(encode(false)).toMatchSnapshot("false");
		expect(encode(new Map([["key", "value"]]))).toMatchSnapshot("map");
		expect(encode({ key: "value" })).toMatchSnapshot("object");
		expect(encode([123, "abc"])).toMatchSnapshot("array");

		const bytes = new Uint8Array([1, 2, 3]);
		expect(encode(bytes)).toMatchSnapshot("uint8array");
		expect(encode(bytes.buffer)).toMatchSnapshot("arraybuffer");
	});

	test("encode/decode", () => {
		const bytes = new Uint8Array([1, 2, 3]);
		const input = {
			posint: 123,
			negint: -123,
			posflo: 123.456,
			negflo: -123.456,
			posbig: POW_2_64 - 1n,
			negbig: -POW_2_64,
			string: "Hello World!",
			undefined: undefined,
			null: null,
			false: false,
			true: true,
			map: new Map([["key", "value"]]),
			array: [123, "abc"],
			uint8array: bytes,
			arraybuffer: bytes.buffer,
		};

		const encoded = encode(input);
		expect(encoded).toMatchSnapshot("encoded input");

		const decoded = decode(encoded);
		expect(decoded).toMatchSnapshot("decoded input");
	});

	describe("infinity", () => {
		test("valid bytes", () => {
			const decoded = decode(
				new Uint8Array([
					95, // infinite bytes start
					65, // byte string, len 1
					1, // Some byte
					66, // byte string, len 2
					1, // Some byte
					2, // Some byte
					255, // break
				]),
			);

			expect(decoded).toMatchObject(new Uint8Array([1, 1, 2]).buffer);
		});

		test("invalid bytes, nested infinite bytes", () => {
			const res = new Promise(() =>
				decode(
					new Uint8Array([
						95, // infinite bytes start
						95, // infinite bytes start (not allowed)
						65, // byte string, len 1
						1, // Some byte
						255, // break
						255, // break
					]),
				),
			);

			expect(res).rejects.toBeInstanceOf(CborRangeError);
		});

		test("invalid bytes, no break", () => {
			const res = new Promise(() =>
				decode(
					new Uint8Array([
						95, // infinite bytes start
						65, // byte string, len 1
						1, // Some byte
						// break (255) is missing
					]),
				),
			);

			expect(res).rejects.toBeInstanceOf(CborRangeError);
		});

		test("invalid bytes, invalid major", () => {
			const res = new Promise(() =>
				decode(
					new Uint8Array([
						95, // infinite bytes start
						96, // text string, len 1 (invalid major)
						1, // Some byte
						255, // break
					]),
				),
			);

			expect(res).rejects.toBeInstanceOf(CborInvalidMajorError);
		});

		test("valid text", () => {
			const decoded = decode(
				new Uint8Array([
					127, // infinite text start
					97, // byte string, len 1
					97, // letter "a"
					98, // byte string, len 2
					98, // letter "b"
					99, // letter "c"
					255, // break
				]),
			);

			expect(decoded).toMatch("abc");
		});

		test("valid array", () => {
			const decoded = decode(
				new Uint8Array([
					159, // infinite array start
					97, // byte string, len 1
					97, // letter "a"
					98, // byte string, len 2
					98, // letter "b"
					99, // letter "c"
					255, // break
				]),
			);

			expect(decoded).toMatchObject(["a", "bc"]);
		});

		test("valid map", () => {
			const decoded = decode(
				new Uint8Array([
					191, // infinite map start
					97, // byte string, len 1
					97, // letter "a"
					98, // byte string, len 2
					98, // letter "b"
					99, // letter "c"
					255, // break
				]),
			);

			expect(decoded).toMatchObject({ a: "bc" });
		});
	});

	describe("partial", () => {
		test("Fails if not enabled", () => {
			const res = new Promise(() => {
				const gap = new Gap();
				encode({ gap });
			});

			expect(res).rejects.toBeInstanceOf(CborPartialDisabled);
		});

		test("Fails to build if fill for gap is missing", () => {
			const res = new Promise(() => {
				const gap = new Gap();
				const partial = encode({ gap }, { partial: true });
				partial.build([]);
			});

			expect(res).rejects.toBeInstanceOf(CborFillMissing);
		});

		describe("Succeeds if configured correctly", () => {
			const name = new Gap<string>();
			const age = new Gap<number>();
			const enabled = new Gap(true);
			const partial = encode(
				[
					"CREATE person SET name = $name, age = $age, enabled = $enabled",
					{
						name,
						age,
						enabled,
					},
				],
				{ partial: true },
			);

			test("with gaps filled", () => {
				const res = decode(partial.build([name.fill("John"), age.fill(30)]));
				expect(res?.[1]).toStrictEqual({
					name: "John",
					age: 30,
					enabled: true,
				});
			});

			test("with defaults overwritten", () => {
				const res = decode(
					partial.build([name.fill("John"), age.fill(30), enabled.fill(false)]),
				);

				expect(res?.[1]).toStrictEqual({
					name: "John",
					age: 30,
					enabled: false,
				});
			});
		});
	});
});
