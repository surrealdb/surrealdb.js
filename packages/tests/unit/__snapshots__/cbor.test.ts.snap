// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`cbor encode basic types: positive integer 1`] = `
Uint8Array [
  24,
  123,
]
`;

exports[`cbor encode basic types: negative integer 1`] = `
Uint8Array [
  56,
  122,
]
`;

exports[`cbor encode basic types: positive float 1`] = `
Uint8Array [
  251,
  64,
  94,
  221,
  47,
  26,
  159,
  190,
  119,
]
`;

exports[`cbor encode basic types: negative float 1`] = `
Uint8Array [
  251,
  192,
  94,
  221,
  47,
  26,
  159,
  190,
  119,
]
`;

exports[`cbor encode basic types: positive bigint 1`] = `
Uint8Array [
  27,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
]
`;

exports[`cbor encode basic types: negative bigint 1`] = `
Uint8Array [
  59,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
]
`;

exports[`cbor encode basic types: string 1`] = `
Uint8Array [
  109,
  72,
  101,
  108,
  108,
  111,
  32,
  10,
  87,
  111,
  114,
  108,
  100,
  33,
]
`;

exports[`cbor encode basic types: undefined 1`] = `
Uint8Array [
  247,
]
`;

exports[`cbor encode basic types: null 1`] = `
Uint8Array [
  246,
]
`;

exports[`cbor encode basic types: true 1`] = `
Uint8Array [
  245,
]
`;

exports[`cbor encode basic types: false 1`] = `
Uint8Array [
  244,
]
`;

exports[`cbor encode basic types: map 1`] = `
Uint8Array [
  161,
  99,
  107,
  101,
  121,
  101,
  118,
  97,
  108,
  117,
  101,
]
`;

exports[`cbor encode basic types: object 1`] = `
Uint8Array [
  161,
  99,
  107,
  101,
  121,
  101,
  118,
  97,
  108,
  117,
  101,
]
`;

exports[`cbor encode basic types: array 1`] = `
Uint8Array [
  130,
  24,
  123,
  99,
  97,
  98,
  99,
]
`;

exports[`cbor encode basic types: uint8array 1`] = `
Uint8Array [
  67,
  1,
  2,
  3,
]
`;

exports[`cbor encode basic types: arraybuffer 1`] = `
Uint8Array [
  67,
  1,
  2,
  3,
]
`;

exports[`cbor encode/decode: encoded input 1`] = `
Uint8Array [
  175,
  102,
  112,
  111,
  115,
  105,
  110,
  116,
  24,
  123,
  102,
  110,
  101,
  103,
  105,
  110,
  116,
  56,
  122,
  102,
  112,
  111,
  115,
  102,
  108,
  111,
  251,
  64,
  94,
  221,
  47,
  26,
  159,
  190,
  119,
  102,
  110,
  101,
  103,
  102,
  108,
  111,
  251,
  192,
  94,
  221,
  47,
  26,
  159,
  190,
  119,
  102,
  112,
  111,
  115,
  98,
  105,
  103,
  27,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  102,
  110,
  101,
  103,
  98,
  105,
  103,
  59,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  255,
  102,
  115,
  116,
  114,
  105,
  110,
  103,
  108,
  72,
  101,
  108,
  108,
  111,
  32,
  87,
  111,
  114,
  108,
  100,
  33,
  105,
  117,
  110,
  100,
  101,
  102,
  105,
  110,
  101,
  100,
  247,
  100,
  110,
  117,
  108,
  108,
  246,
  101,
  102,
  97,
  108,
  115,
  101,
  244,
  100,
  116,
  114,
  117,
  101,
  245,
  99,
  109,
  97,
  112,
  161,
  99,
  107,
  101,
  121,
  101,
  118,
  97,
  108,
  117,
  101,
  101,
  97,
  114,
  114,
  97,
  121,
  130,
  24,
  123,
  99,
  97,
  98,
  99,
  106,
  117,
  105,
  110,
  116,
  56,
  97,
  114,
  114,
  97,
  121,
  67,
  1,
  2,
  3,
  107,
  97,
  114,
  114,
  97,
  121,
  98,
  117,
  102,
  102,
  101,
  114,
  67,
  1,
  2,
  3,
]
`;

exports[`cbor encode/decode: decoded input 1`] = `
{
  "array": [
    123,
    "abc",
  ],
  "arraybuffer": ArrayBuffer [
    1,
    2,
    3,
  ],
  "false": false,
  "map": {
    "key": "value",
  },
  "negbig": -18446744073709551616n,
  "negflo": -123.456,
  "negint": -123,
  "null": null,
  "posbig": 18446744073709551615n,
  "posflo": 123.456,
  "posint": 123,
  "string": "Hello World!",
  "true": true,
  "uint8array": ArrayBuffer [
    1,
    2,
    3,
  ],
  "undefined": undefined,
}
`;
